File: src/frontend/inventario/lib/app/data/model/inventario_contagem_det_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:inventario/app/data/model/model_imports.dart';
import 'package:inventario/app/data/domain/domain_imports.dart';

class InventarioContagemDetModel {
	int? id;
	int? idInventarioContagemCab;
	int? idProduto;
	double? contagem01;
	double? contagem02;
	double? contagem03;
	String? fechadoContagem;
	double? quantidadeSistema;
	double? acuracidade;
	double? divergencia;
	ProdutoModel? produtoModel;

	InventarioContagemDetModel({
		this.id,
		this.idInventarioContagemCab,
		this.idProduto,
		this.contagem01,
		this.contagem02,
		this.contagem03,
		this.fechadoContagem,
		this.quantidadeSistema,
		this.acuracidade,
		this.divergencia,
		this.produtoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'contagem01',
		'contagem02',
		'contagem03',
		'fechado_contagem',
		'quantidade_sistema',
		'acuracidade',
		'divergencia',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Contagem01',
		'Contagem02',
		'Contagem03',
		'Fechado Contagem',
		'Quantidade Sistema',
		'Acuracidade',
		'Divergencia',
	];

	InventarioContagemDetModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idInventarioContagemCab = jsonData['idInventarioContagemCab'];
		idProduto = jsonData['idProduto'];
		contagem01 = jsonData['contagem01']?.toDouble();
		contagem02 = jsonData['contagem02']?.toDouble();
		contagem03 = jsonData['contagem03']?.toDouble();
		fechadoContagem = InventarioContagemDetDomain.getFechadoContagem(jsonData['fechadoContagem']);
		quantidadeSistema = jsonData['quantidadeSistema']?.toDouble();
		acuracidade = jsonData['acuracidade']?.toDouble();
		divergencia = jsonData['divergencia']?.toDouble();
		produtoModel = jsonData['produtoModel'] == null ? ProdutoModel() : ProdutoModel.fromJson(jsonData['produtoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idInventarioContagemCab'] = idInventarioContagemCab != 0 ? idInventarioContagemCab : null;
		jsonData['idProduto'] = idProduto != 0 ? idProduto : null;
		jsonData['contagem01'] = contagem01;
		jsonData['contagem02'] = contagem02;
		jsonData['contagem03'] = contagem03;
		jsonData['fechadoContagem'] = InventarioContagemDetDomain.setFechadoContagem(fechadoContagem);
		jsonData['quantidadeSistema'] = quantidadeSistema;
		jsonData['acuracidade'] = acuracidade;
		jsonData['divergencia'] = divergencia;
		jsonData['produtoModel'] = produtoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idInventarioContagemCab = plutoRow.cells['idInventarioContagemCab']?.value;
		idProduto = plutoRow.cells['idProduto']?.value;
		contagem01 = plutoRow.cells['contagem01']?.value?.toDouble();
		contagem02 = plutoRow.cells['contagem02']?.value?.toDouble();
		contagem03 = plutoRow.cells['contagem03']?.value?.toDouble();
		fechadoContagem = plutoRow.cells['fechadoContagem']?.value != '' ? plutoRow.cells['fechadoContagem']?.value : '01';
		quantidadeSistema = plutoRow.cells['quantidadeSistema']?.value?.toDouble();
		acuracidade = plutoRow.cells['acuracidade']?.value?.toDouble();
		divergencia = plutoRow.cells['divergencia']?.value?.toDouble();
		produtoModel = ProdutoModel();
		produtoModel?.nome = plutoRow.cells['produtoModel']?.value;
	}	

	InventarioContagemDetModel clone() {
		return InventarioContagemDetModel(
			id: id,
			idInventarioContagemCab: idInventarioContagemCab,
			idProduto: idProduto,
			contagem01: contagem01,
			contagem02: contagem02,
			contagem03: contagem03,
			fechadoContagem: fechadoContagem,
			quantidadeSistema: quantidadeSistema,
			acuracidade: acuracidade,
			divergencia: divergencia,
		);			
	}

	
}

File: src/backend/inventario/src/model/inventario_contagem_det_model.py
from src import db
from src.model.produto_model import ProdutoModel


class InventarioContagemDetModel(db.Model):
    __tablename__ = 'inventario_contagem_det'

    id = db.Column(db.Integer, primary_key=True)
    contagem01 = db.Column(db.Float)
    contagem02 = db.Column(db.Float)
    contagem03 = db.Column(db.Float)
    fechado_contagem = db.Column(db.String(2))
    quantidade_sistema = db.Column(db.Float)
    acuracidade = db.Column(db.Float)
    divergencia = db.Column(db.Float)
    id_inventario_contagem_cab = db.Column(db.Integer, db.ForeignKey('inventario_contagem_cab.id'))
    id_produto = db.Column(db.Integer, db.ForeignKey('produto.id'))

    produto_model = db.relationship('ProdutoModel', foreign_keys=[id_produto])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_inventario_contagem_cab = data.get('idInventarioContagemCab')
        self.id_produto = data.get('idProduto')
        self.contagem01 = data.get('contagem01')
        self.contagem02 = data.get('contagem02')
        self.contagem03 = data.get('contagem03')
        self.fechado_contagem = data.get('fechadoContagem')
        self.quantidade_sistema = data.get('quantidadeSistema')
        self.acuracidade = data.get('acuracidade')
        self.divergencia = data.get('divergencia')

    def serialize(self):
        return {
            'id': self.id,
            'idInventarioContagemCab': self.id_inventario_contagem_cab,
            'idProduto': self.id_produto,
            'contagem01': self.contagem01,
            'contagem02': self.contagem02,
            'contagem03': self.contagem03,
            'fechadoContagem': self.fechado_contagem,
            'quantidadeSistema': self.quantidade_sistema,
            'acuracidade': self.acuracidade,
            'divergencia': self.divergencia,
            'produtoModel': self.produto_model.serialize() if self.produto_model else None,
        }